src/java/org/apache/cassandra/streaming/async/StreamCompressionSerializer.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.cassandra.utils.ChecksumType;
+
+/**
+ * A serialiazer for stream compressed files (see package-level documentation). Much like a typical compressed
+ * output stream, this class operates on buffers or chunks of the data at a a time. The format for each compressed
+ * chunk is as follows:
+ *
+ * - int - compressed payload length
+ * - int - uncompressed payload length
+ * - int - checksum of lengths
+ * - bytes - compressed payload
+ * - int - checksum of uncompressed data
+ *
+ * Note: instances are not thread-safe.
+ */
+class StreamCompressionSerializer
+{
+    /**
+     * Length of heaer data, which includes compressed length, uncompressed length, the their chunksum.
+     */
+    private static final int HEADER_LENGTH = 12;
+
+    enum DeserializeState { LENGTHS, PAYLOAD }
+
+    private final LZ4Compressor compressor;
+    private final LZ4FastDecompressor decompressor;
+
+    private DeserializeState deserializeState;
+    private ChunkLengths chunkLengths;
+    private final ByteBuffer lengthChecksumBuffer = ByteBuffer.allocate(8);
+
+    StreamCompressionSerializer(LZ4Compressor compressor, LZ4FastDecompressor decompressor)
+    {
+        this.compressor = compressor;
+        this.decompressor = decompressor;
+        deserializeState = DeserializeState.LENGTHS;
+    }
+
+    public ByteBuf serialize(ByteBuf in, int version) throws IOException
+    {
+        final ChecksumType checksum = ChecksumType.CRC32;
+        final int uncompressedLength = in.readableBytes();
+
+        // Note: there's better alternatives to creating extra buffers and memcpy'ing things,
+        // but there's a bug in the lz4-java 1.3.0 jar, and it's unmaintained atm :(
+        byte[] inBuffer = new byte[uncompressedLength];
+        in.readBytes(inBuffer);
+        int payloadChecksum = (int)checksum.of(inBuffer, 0, uncompressedLength);
+
+        byte[] outBuffer = new byte[compressor.maxCompressedLength(uncompressedLength) + HEADER_LENGTH + 4];
+        final int compressedLength = compressor.compress(inBuffer, 0, uncompressedLength, outBuffer, HEADER_LENGTH);
+
+        ByteBuf out = Unpooled.wrappedBuffer(outBuffer, 0, compressedLength + HEADER_LENGTH + 4);
+        out.writerIndex(0);
+        out.writeInt(compressedLength);
+        out.writeInt(uncompressedLength);
+        out.writeInt(checksumLengths(checksum, compressedLength, uncompressedLength));
+        out.writerIndex(HEADER_LENGTH + compressedLength);
+        out.writeInt(payloadChecksum);
+        return out;
+    }
+
+    int checksumLengths(ChecksumType checksum, int compressedLength, int uncompressedLength)
+    {
+        lengthChecksumBuffer.clear();
+        lengthChecksumBuffer.putInt(0, compressedLength);
+        lengthChecksumBuffer.putInt(4, uncompressedLength);
+        return (int) checksum.of(lengthChecksumBuffer);
+    }
+
+    public ByteBuf deserialize(AppendingByteBufInputPlus in, int version) throws IOException
```

Comment:
It's not available that blocks. It's reading a value out of the stream. So you think the number of bytes you need is 100, you go to deserialize the message and it tries to deserialize 101 bytes due to a bug. The netty thread now blocks waiting on itself to retrieve more bytes.